File: website/src/cad/feature-dsl-capabilities.ts
export const LITECAD_FEATURE_DSL_CAPABILITY_REGISTRY = {
  version: 1,
  features: [
    'sketch',
    'box',
    'box_cut',
    'extrude',
    'extrude_cut',
    'cylinder',
    'cylinder_cut',
    'sphere',
    'ellipsoid',
    'ellipse_extrude',
    'tapered_extrude',
    'revolve',
    'sweep',
    'loft',
    'fillet',
    'chamfer',
    'boolean',
  ],
  booleanOperations: ['union', 'subtract', 'intersect'],
  sketchPlanes: ['XY', 'XZ', 'YZ'],
} as const

const supportedFeatureDSLTypes = new Set<string>(LITECAD_FEATURE_DSL_CAPABILITY_REGISTRY.features)

export function isSupportedFeatureDSLType(type: unknown): type is (typeof LITECAD_FEATURE_DSL_CAPABILITY_REGISTRY.features)[number] {
  return typeof type === 'string' && supportedFeatureDSLTypes.has(type)
}

export function assertFeatureDSLCompilerCoverage(compilerTypes: readonly string[]) {
  const compilerTypeSet = new Set(compilerTypes)
  const missing = LITECAD_FEATURE_DSL_CAPABILITY_REGISTRY.features.filter((type) => !compilerTypeSet.has(type))
  const extra = compilerTypes.filter((type) => !supportedFeatureDSLTypes.has(type))
  if (missing.length === 0 && extra.length === 0) {
    return
  }
  const details = [
    missing.length > 0 ? `missing ${missing.join(', ')}` : '',
    extra.length > 0 ? `extra ${extra.join(', ')}` : '',
  ].filter(Boolean)
  throw new Error(`Feature DSL compiler coverage mismatch: ${details.join('; ')}`)
}

File: internal/service/feature_dsl_capabilities.go
package service

import "strings"

// LiteCADFeatureDSLCapabilityRegistry describes the backend-owned generated-source contract.
type LiteCADFeatureDSLCapabilityRegistry struct {
	Features          []string
	BooleanOperations []string
	SketchPlanes      []string
}

var liteCADFeatureDSLCapabilityRegistry = LiteCADFeatureDSLCapabilityRegistry{
	Features: []string{
		"sketch", "box", "box_cut", "extrude", "extrude_cut", "cylinder", "cylinder_cut", "sphere",
		"ellipsoid", "ellipse_extrude", "tapered_extrude", "revolve", "sweep", "loft", "fillet", "chamfer", "boolean",
	},
	BooleanOperations: []string{"union", "subtract", "intersect"},
	SketchPlanes:      []string{"XY", "XZ", "YZ"},
}

// LiteCADFeatureDSLCapabilities returns an immutable snapshot of generated-source capabilities.
func LiteCADFeatureDSLCapabilities() LiteCADFeatureDSLCapabilityRegistry {
	return LiteCADFeatureDSLCapabilityRegistry{
		Features:          append([]string(nil), liteCADFeatureDSLCapabilityRegistry.Features...),
		BooleanOperations: append([]string(nil), liteCADFeatureDSLCapabilityRegistry.BooleanOperations...),
		SketchPlanes:      append([]string(nil), liteCADFeatureDSLCapabilityRegistry.SketchPlanes...),
	}
}

func isLiteCADFeatureDSLFeatureType(featureType string) bool {
	for _, supportedType := range liteCADFeatureDSLCapabilityRegistry.Features {
		if featureType == supportedType {
			return true
		}
	}
	return false
}

func replaceCapabilityList(source, prefix, suffix string) string {
	start := strings.Index(source, prefix)
	if start < 0 {
		return source
	}
	listStart := start + len(prefix)
	endOffset := strings.Index(source[listStart:], suffix)
	if endOffset < 0 {
		return source
	}
	return source[:listStart] + strings.Join(liteCADFeatureDSLCapabilityRegistry.Features, ", ") + source[listStart+endOffset:]
}
